[PR] 주문 단건 조회 기능 구현

qkrtkdwns3410 opened the pull request:
1. 목표
- 주문 단건 조회 관련 기능 구현

2. 내용
- 주문 단순 조회 기능
- 사용자 역할별 주문 조회 가능여부에 대해 분기처리하는 auth service 생성
- 서비스, 매퍼 성공 실패 테스트 코드 작성 
- + 2024.01.31
	- 컨트롤러 퍼시스턴트 영역 테스트 코드 작성 완료
- 2024.02.02
    - 서비스단의 인증로직을 컨트롤러에서 분기처리하는 것으로 변경했습니다.
- [리팩토링] 별도 Exception 쪽 코드에 대해 리팩토링을 수행했습니다
- 주문 단건 조회시 인증 실패 에러 코드를 추가하려고했으나, 불필요하게 Controller Advice 가 두꺼워 지는 것 같아서 수정했습니다.
- [리팩토링] order api 에 대해 복수형 리소스로 엔드포인트를 변경했습니다.
- [리팩토링] OrderRequestDTO 에서 shownDeletedOrder 에 대해 null 값을 주는 경우를 위하여 래퍼 타입으로 변경했습니다.
- [추가] 멤버 도메인의 static helper method 를 추가했습니다.

https://www.figma.com/file/XuopIJakOLO1ffjsMgqQKG/Order-Flow?type=whiteboard&node-id=0%3A1&t=xVzVmjkDbl6eItib-1

qkrtkdwns3410 committed:
feat: Working on the single order inquiry service-related task

qkrtkdwns3410 committed:
test: Add unit test for normal inquiry of individual orders

dev-mentoring commented on src/test/java/com/psj/itembrowser/order/service/OrderSelectServiceTest.java:
```diff
@@ -0,0 +1,68 @@
+    void When_GetOrder_Expect_ThrowNotFoundException() {
+        //given
+        given(orderPersistence.getOrder(validOrderRequestDTO)).willReturn(validOrder);
+        //when
```
개행 부탁드립니다.

qkrtkdwns3410 replied:
수정했습니다!

dev-mentoring commented on src/main/java/com/psj/itembrowser/order/controller/OrderApiController.java:
```diff
@@ -1,21 +1,34 @@
 @RestController
+@Slf4j
 @RequiredArgsConstructor
 @RequestMapping("/v1/api/order")
```
복수로 표현 부탁드립니다. 
```java
/v1/api/orders
```

dev-mentoring replied:
추가적으로 v1과 같은 버전관리가 Controller 상단에 위치하면 나중에 버전관리하기가 어려울겁니다. 각 컨트롤러 API에서 작성하시는게 버전관리가 쉬울겁니다.

qkrtkdwns3410 replied:
복수 표현 및 버전관리를 메서드 개별로 각각 선언으로 변경했습니다.

dev-mentoring commented on src/main/java/com/psj/itembrowser/order/domain/dto/request/OrderRequestDTO.java:
```diff
@@ -1,15 +1,25 @@
 public class OrderRequestDTO {
+    
     Long id;
     boolean shownDeletedOrder;
```
이렇게 필드값으로 관리하는 것이 좋아보이지는 않는데 다른 방법 없을지 고민됩니다.

dev-mentoring commented on src/test/java/com/psj/itembrowser/order/service/OrderSelectServiceTest.java:
```diff
@@ -0,0 +1,68 @@
+        //when
+        OrderResponseDTO result = orderService.getOrder(validOrderRequestDTO);
+        
+        //then
+        assertThat(result.getId()).isEqualTo(validOrder.getId());
+        assertThat(result.getOrdererId()).isEqualTo(validOrder.getOrdererId());
+        assertThat(result.getOrderStatus()).isEqualTo(validOrder.getOrderStatus());
+        assertThat(result.getPaidDate()).isEqualTo(validOrder.getPaidDate());
+        assertThat(result.getShippingInfoId()).isEqualTo(validOrder.getShippingInfoId());
+        assertThat(result.getCreatedDate()).isEqualTo(validOrder.getCreatedDate());
+        assertThat(result.getUpdatedDate()).isEqualTo(validOrder.getUpdatedDate());
+        assertThat(result.getDeletedDate()).isEqualTo(validOrder.getDeletedDate());
```
assertThat 다음에 비교하는 비교군은 실 데이터 값이 들어가야합니다. given에서 만들어낸 변수로 넣게되면 비교하는 의미가 then쪽에서 없을 때가 많고 코드를 읽을때도 수정하는 부분에서 눈에 띄기가 쉽지 않습니다.

dev-mentoring reviewed (changes requested):
좋은 코드 감사합니다!

코드리뷰 남겨드립니다!

qkrtkdwns3410 committed:
feat: Modified Member's Role

        ROLE_CUSTOMER("일반 구매자"),
        ROLE_STORE_SELLER("상점 판매자"),
        ROLE_ADMIN("관리자");

qkrtkdwns3410 committed:
refactor: update API endpoints from 'order' to 'orders'

- Rename '/v1/api/order' to '/v1/api/orders' in OrderApiController
- Update all associated tests in OrderDeleteApiControllerTest
- Follow REST API design convention of using plural resource names

qkrtkdwns3410 committed:
style: line break

qkrtkdwns3410 committed:
refactor: add role check method

qkrtkdwns3410 committed:
refactor: add static method in Member class and add static method to dto

qkrtkdwns3410 committed:
refactor: The steps of the exception class were divided and simplified.

qkrtkdwns3410 committed:
feat: add error code and advice

qkrtkdwns3410 committed:
feat: Injection of authentication service and completion of logic related to single order inquiry

qkrtkdwns3410 committed:
test: modify the test code to operate normally

qkrtkdwns3410 committed:
test: Implementation of test code for single order inquiry

qkrtkdwns3410 committed:
refactor: ordererId to ordererNo in testing and domain

qkrtkdwns3410 committed:
test: write mapper test code

qkrtkdwns3410 committed:
refactor: inversion for real data and mocked responses

qkrtkdwns3410 committed:
refactor: Added a static method to the value type in the member aggregate.

qkrtkdwns3410 committed:
refactor: Remove Builder

qkrtkdwns3410 committed:
test: Write persistence area test code

qkrtkdwns3410 committed:
refactor: Changed the controller API to a separate URL notation for each endpoint for versioning.

qkrtkdwns3410 committed:
refactor: Update for api versioning

ka-MS commented on src/main/java/com/psj/itembrowser/common/handler/ControllerAdvice.java:
```diff
@@ -1,73 +1,56 @@
-        return new ResponseEntity<>(apiError, e.getErrorCode().getStatus());
-    }
+	
+	@ExceptionHandler(CustomRuntimeException.class)
```
@ExceptionHandler(BadRequestException.class, IllegalStateException.class)
이런식으로 핸들러에 나열하는 방식은 어떤가요?

qkrtkdwns3410 replied:
저는 커스텀 에러코드를 핸들러에 전달하기위해서 별도 CustomRuntimeException 을 만들어서 사용했었습니다.
저 부분에서 만약에 BadRequestException.class, IllegalStateException.class 을 받으면 매개변수로 어떤 에러를 넣어줄것인지가 애매해지는데.. 혹시 그 부분은 어떻게 생각하시는지요

ka-MS commented on src/main/java/com/psj/itembrowser/common/handler/ControllerAdvice.java:
```diff
@@ -1,73 +1,56 @@
+	@ExceptionHandler(CustomRuntimeException.class)
+	public final ResponseEntity<Object> notFoundExceptionHandler(CustomRuntimeException e,
+		WebRequest request) {
+		ApiError apiError = new ApiError(
+			e.getErrorCode().getStatus()
+				.value(),
+			e.getErrorCode().getMessage(),
+			e.getMessage(),
+			request.getDescription(false)
+		);
+		
+		return new ResponseEntity<>(apiError, e.getErrorCode().getStatus());
```
동일한 코드가 반복되는데 메소드로 빼는건 어떨까요?

qkrtkdwns3410 replied:
```

	@ExceptionHandler(CustomRuntimeException.class)
	public final ResponseEntity<Object> notFoundExceptionHandler(CustomRuntimeException e,
		WebRequest request) {
		ApiError apiError = new ApiError(
			e.getErrorCode().getStatus()
				.value(),
			e.getErrorCode().getMessage(),
			e.getMessage(),
			request.getDescription(false)
		);
		
		return new ResponseEntity<>(apiError, e.getErrorCode().getStatus());
	}
	
	@ExceptionHandler(CustomIllegalStateException.class)
	public final ResponseEntity<Object> badRequestExceptionHandler(CustomIllegalStateException e,
		WebRequest request) {
		ApiError apiError = new ApiError(
			e.getErrorCode().getStatus()
				.value(),
			e.getErrorCode().getMessage(),
			e.getMessage(),
			request.getDescription(false)
		);
		
		return new ResponseEntity<>(apiError, e.getErrorCode().getStatus());
	}
	
	@ExceptionHandler(CustomAuthenticationException.class)
	public final ResponseEntity<Object> authenticationExceptionHandler(CustomAuthenticationException e,
		WebRequest request) {
		ApiError apiError = getApiError(e, request);
		
		return new ResponseEntity<>(apiError, e.getErrorCode().getStatus());
	}
	
	private static ApiError getApiError(CustomAuthenticationException e, WebRequest request) {
		return new ApiError(
			e.getErrorCode().getStatus()
				.value(),
			e.getErrorCode().getMessage(),
			e.getMessage(),
			request.getDescription(false)
		);

```

저 부분도 지금 매개변수로 받는 exception 이 다르다보니 메서드로 분리하기가 조금 까다로운 것 같습니다.
혹시 어떤식으로 해야할지 잘 모르겠네요..

각 커스텀 excpetion 이 상속받는 exception 다르다보니..

dev-mentoring replied:
우선 이 부분에 대해서는 따로 빼지 않는 것이 좋아보입니다. 

하나의 메서드안에서 처리가 n번 반복이 된다면 메서드나 클래스로 빼는 것을 고려해볼 수 있습니다.

이 경우, 각각의 핸들러에서 처리가 되는 로직이므로 반복으로 보기 어렵습니다.

ka-MS commented on src/main/java/com/psj/itembrowser/order/domain/dto/request/OrderRequestDTO.java:
```diff
@@ -1,15 +1,29 @@
-    Long id;
-    boolean shownDeletedOrder;
+	
+	Long id;
+	Boolean shownDeletedOrder = false;
```
필드의 접근제한자가 private이 아닌 defalut인 이유가 있을까요?

qkrtkdwns3410 replied:
해당 부분은 누락으로 수정했습니다 감사합니다

ka-MS reviewed (commented):
고생하셨습니다! 코드 리뷰 쉽지 않네요

qkrtkdwns3410 committed:
test: Update OrderSelectApiControllerTest to include tests for order retrieval and error handling.

qkrtkdwns3410 committed:
refactor: Declare private access limiter for orderrequestdto

qkrtkdwns3410 committed:
remove: Remove comment from OrderMapper

dev-mentoring commented on src/main/java/com/psj/itembrowser/order/controller/OrderApiController.java:
```diff
@@ -1,21 +1,36 @@
+    public ResponseEntity<OrderResponseDTO> getOrder(@PathVariable Long orderId) {
+        OrderRequestDTO orderRequestDTO = OrderRequestDTO.forActiveOrder(orderId);
+        return ResponseEntity.ok(orderService.getOrder(orderRequestDTO));
+    }
```
@RequestParam 을 이용해서 하셔도 좋고 권한에 따라서 결과가 완전히 분리되어야 한다면

기존에
```java
private Boolean shownDeletedOrder = false;
```
위의 값을 필드로 DTO에서 관리하고 있는데,
여기서  Controller에서 권한 분기를 처리하시고 그에 맞게 Service 로직을 각각 호출해서 처리하면 더 좋아보입니다.

dev-mentoring commented on src/main/java/com/psj/itembrowser/order/domain/dto/request/OrderRequestDTO.java:
```diff
@@ -1,15 +1,29 @@
+	
+	public static OrderRequestDTO forAllOrder(Long id) {
+		return new OrderRequestDTO(id, null);
+	}
```
이 부분도 수정이 될 것 같습니다. 불필요해서 삭제 될 것을 기대합니다.

dev-mentoring commented on src/main/java/com/psj/itembrowser/order/domain/vo/Order.java:
```diff
@@ -27,14 +27,14 @@ public class Order implements Cancelable {
         Order order = new Order();
         order.id = id;
-        order.ordererId = ordererId;
+        order.ordererId = ordererNo;
```
이 부분 체크 한번 해보시고, No 말고 다른 표현을 적으시면 좋겠습니다.

dev-mentoring commented on src/main/java/com/psj/itembrowser/order/persistence/OrderPersistence.java:
```diff
@@ -46,4 +43,12 @@ public Order findOrderStatusForUpdate(long orderId) {
+        if (findOrder == null) {
+            throw new NotFoundException(ORDER_NOT_FOUND);
+        }
+        return findOrder;
```
if문 전과 return 전에 개행 부탁드립니다.

dev-mentoring commented on src/main/java/com/psj/itembrowser/security/service/impl/UserDetailsServiceImpl.java:
```diff
@@ -1,93 +1,104 @@
+		if (memberResponseDTO == null) {
+			throw new UsernameNotFoundException("username " + email + " is not found");
+		}
+		return new CustomUserDetails(memberResponseDTO);
```
개행 이 부분 좀 더 신경써서 해주세요!

dev-mentoring commented on src/main/java/com/psj/itembrowser/security/service/impl/UserDetailsServiceImpl.java:
```diff
@@ -1,93 +1,104 @@
+	@Getter
+	@RequiredArgsConstructor
+	public static final class CustomUserDetails implements UserDetails {
+		//QUESTION 스프링 예제에서는 상속을 받았지만, 상속받는 이유를 잘 모르겠다.
```
주석 불필요한 부분이면 삭제해주세요

dev-mentoring reviewed (changes requested):
수고하셨습니다 ! 요청드린 부분 수정 부탁드릴게요!

qkrtkdwns3410 committed:
refactor: Progressive MockDataGen class deletion

qkrtkdwns3410 committed:
remove: Remove old member insert sql

qkrtkdwns3410 committed:
refactor: Change orderer ID to orderer number across multiple files

qkrtkdwns3410 committed:
chore: Remove Comment

qkrtkdwns3410 committed:
refactor: Fixed exception code

qkrtkdwns3410 committed:
chore: Simple variable name change

qkrtkdwns3410 committed:
refactor: Changed method to make if conditional statement easier to understand

qkrtkdwns3410 committed:
remove: Delete unused methods

qkrtkdwns3410 committed:
fix: While resolving issues related to member information authentication, the overall logic needs to be modified due to NPE occurring in the logic.

qkrtkdwns3410 committed:
feat: Add member response to OrderResponseDTO

qkrtkdwns3410 committed:
remove: The seller's conditions were decided to be filtered at the controller level.

qkrtkdwns3410 committed:
feat: Set member information to be searched in order-mapper.xml

qkrtkdwns3410 committed:
feat: Access modified to PROTECTED

qkrtkdwns3410 committed:
remove: Deleted due to change in authentication-related logic

qkrtkdwns3410 committed:
test: Refactor test cases for order retrieval functionality

qkrtkdwns3410 committed:
remove: Deleted Class due to change in authentication method

qkrtkdwns3410 committed:
refactor: Modificate overall logic due to change in authentication method Change from controller stage to branching based on authentication

qkrtkdwns3410 committed:
fix: Receive separate authentication from the controller cause A User object is required to

qkrtkdwns3410 committed:
feat: Set not to return a stacktrace in the response.

qkrtkdwns3410 committed:
feat: Create a custom annotation that wraps @AuthenticationPrincipal

qkrtkdwns3410 committed:
fix: Changed the authentication method to JwtAuthenticationToken. The existing UsernamePasswordAuthenticationToken does not match the JWT method.

qkrtkdwns3410 committed:
fix: Changes in test code due to changes in authentication method

qkrtkdwns3410 committed:
fix: Excluded passwords from serialization

qkrtkdwns3410 committed:
refactor: Changed to perform only the member role within the member , In accordance with DDD
